### ISSUE TITLE
Watcher not working using TypeScript?

### ISSUE THREAD

**Issue by kaosdev** [jlmakes/karma-rollup-preprocessor#72]
Seems like that karma auto watch break when using `karma-rollup-preprocessor`.
The tests are run when i change something, but with no effect.
Indeed if I make a test fail, I see karma running but it will state that tests passes.

It may be a configuration problem, so here I made a small reproduction of the issue https://github.com/kaosdev/karma-rollup-preprocessor-watch-issue

I saw there are other issue related to auto watch not working, but this seems different, and as far as I understand this feature changed recently (maybe with 7.0.4?).

**Comment by jlmakes** [comment]
Thanks for the reproduction. I removed TypeScript from the project and everything works normally. That narrows things.

**Comment by kaosdev** [comment]
Can confirm that, using babel watch works correctly. After digging a while I found that `@rollup/plugin-typescript` uses an internal watcher, maybe this is causing some errors?

**Comment by jlmakes** [comment]
`karma-rollup-preprocessor` has its own watcher that manually re-runs Karma, but presently I don't deeply understand how Karma goes about generating/serving temporary files. Although, if it works without the TS plugin, it makes me think the problem lies in how `@rollup/plugin-typescript` names its output files, perhaps related to the file watcher or a caching strategy. 🤔

**Comment by jlmakes** [comment]
`karma-rollup-preprocessor` has its own watcher that manually re-runs Karma, but presently I don't deeply understand how Karma goes about generating/serving temporary files. Although, if it works without the TS plugin, it makes me think the problem lies in how `@rollup/plugin-typescript` names its output files, perhaps related to the file watcher or a caching strategy. 🤔

There is a bit of file name shuffling in order to get source maps working, and to account for file extension changes. If there is a fix that can be implemented in preprocessor land (as opposed to in the plug-in itself) it likely has to do with that.

There are other users that have been using TypeScript though: https://github.com/jlmakes/karma-rollup-preprocessor/blob/master/CHANGELOG.md#fixed-6 but it looks like it was done using Babel.

**Comment by MMMalik** [comment]
I can confirm that there is an issue with watching files and official TS plugin `@rollup/plugin-typescript`. Switching to `@rollup/plugin-babel` helped in my case:

```
plugins: [
	progress(),
	babel( {
		babelHelpers: 'bundled',
		presets: [
			'@babel/preset-env',
			'@babel/preset-typescript',
			'@babel/preset-react'
		],
		extensions: [ '.ts', '.tsx', '.js' ],
		exclude: 'node_modules/**'
	} ),
	nodeResolve( {
		preferBuiltins: false,
		extensions: [ '.ts', '.tsx', '.js' ]
	} ),
	commonJs(),
	polyfillNode()
],
```

With babel plugin it's just transpilation (no type checking) but I don't think it's a big deal in case of generating test bundle.

**Comment by xenobytezero** [comment]
So I have been spending a few days looking at this and it looks like it's partially related to `@rollup/typescript` and partially down to the structure of this preprocessor.

After diving through the `@rollup/typescript` code, it looks like the plugins "emitted files" cache is maintained at the plugin instance level, not per Rollup run. Since this preprocessor calls `rollup.rollup()` multiple times with the same config, the instance of the plugin is maintained as does the cache, hence every time the file changes the previously compiled one is generated again.

I assume something different happens when running `rollup.watch()` as my main build (using an identical rollup config) works fine. 

Is there a way to modify the preprocessor to use `rollup.watch()` and its internal watcher, while still saying to Karma "hey stuff has changed, you should do what you need to" in the same way as `refreshFiles` does?

I'm going to keep messing around with this on my end to see if I can make it work, but any pointers would be great.

**Comment by jlmakes** [comment]
@xenobytezero I'd be curious to see your changes. If you have the time, please submit a PR.

**Comment by xenobytezero** [comment]
The changes are tied up in code it's a little tricky to open source right now, will see what I can do.

**Comment by YonatanKra** [comment]
@jlmakes I made it work. 
I can't find the logic in the tests in order to write tests for it. 
How do the tests work? All I see is a bunch of modules that import each other.

**Comment by YonatanKra** [comment]
@jlmakes - do you think you could integrate this into the package?

**Comment by jlmakes** [comment]
@YonatanKra I'll cut a new release shortly!

**Comment by mattalexx** [comment]
Just spend an hour solving this. My mistake for not looking further after noticing that my version matched your package.json at master. 

Are you going to release soon?

**Comment by jlmakes** [comment]
Thanks for your patience, the fix landed in `7.0.8` @mattalexx @YonatanKra

**Comment by mattalexx** [comment]
Thanks @jlmakes 

cc @Gusted
